### PR TITLE
Refactor Movie Rental System (Strategy Pattern & Modularization)

### PR DESCRIPTION
**Pull request by nicksan222**
## Overview
This PR refactors the movie rental system by implementing the **Strategy Design Pattern** for pricing logic and modularizing the codebase for better clarity and maintainability. No breaking changes were introduced.

## 🎯 Key Changes
- **Strategy Pattern Implementation:**
  - Replaced the previous `PriceCode` system with a flexible, strategy-based model (`RegularPricingStrategy`, `ChildrensPricingStrategy`, and `NewReleasePricingStrategy`).
  - Updated `Movie` class to utilize strategy instances directly.
  - Deprecated static constants (`PRICE_CODE_*`) in favor of strategy objects.

- **Codebase Modularization:**
  - Split the previous monolithic `engine.ts` into clear, purpose-driven modules (`store.ts`, `movie.ts`, `customer.ts`, `rental.ts`).

- **Enhanced Testing:**
  - Maintained **complete test coverage** (now is 97% but just for the lack of coverage of fixed value private getters in pricing classes).
  - Existing tests were preserved as-is and simply moved to align with the new folder structure.
  - Added new targeted unit tests for each pricing strategy to address specific edge cases and ensure full coverage of pricing behaviors.

## 📋 Notes for Reviewer(s)
Some important considerations guided the preparation of this PR

| **Topic** | **Details** |
|-----------|-------------|
| **1. Preservation of Existing Behavior** | - Current implementation has **100% test coverage**.<br>- All existing behaviors are preserved and must remain unchanged. |
| **2. Avoidance of Breaking Changes** | - Existing usage patterns (as shown in `index.ts`) **must remain functional** without requiring changes. |
| **3. Pricing Strategy Enforcement** | - Follows a model: **base days + cost per extra day**.<br>- While clear and consistent, it may be a limitation in future real-world use cases.<br>- Worth discussing based on actual business needs. |
| **4. Intentional Omission of Input Validation** | - **No input validation** included to maintain backward compatibility.<br>- Adding validation would introduce **new behaviors**, which is intentionally avoided. |
| **5. Styling Guidelines** | - Since PR templates, branch naming conventions, and comment styles are neither explicitly defined nor clearly inferable from the codebase, they have been established based on common best practices. |

## ✅ Checklist for Future Improvements

| **Improvement** | **Description** | **Is Breaking?** |
|------------------|------------------|------------------|
| Refactor `statement()` method | The current method name (statement) is vague and not descriptive. Consider renaming it (e.g., generateRentalSummary()) and potentially using a Strategy pattern for future multi-format support. Not done yet to avoid unnecessary changes until a new feature explicitly requires it. | ❌ (if kept also the old method as deprecated) |
| Add input validation | Introduce validation logic to ensure safer execution and allow for potential idempotency handling. | ✅ (if strict enforcement affects current consumers) |
| Remove deprecated elements | Gradually eliminate deprecated constants and interfaces as the system evolves. | ✅ |

### CODE REVIEW

File: src/engine/index.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export * from './store'
+export * from './pricing'
```
Comment (nicksan222) [inline src/engine/index.ts:2]: Exporting the pricing classes allows better inversion of control, since the consumers of the store class can directly pass instances of a `PricingStrategy` directly instead of relying on store-defined static constants, as well as creating more complex calculations functions extending the base pricing class.

File: src/engine/store.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import {
+    ChildrensPricingStrategy,
+    RegularPricingStrategy,
+    NewReleasePricingStrategy,
+    PricingStrategy
+} from "./pricing"
+
+export class Store {
+    /** @deprecated Use `new RegularPricingStrategy()` instead */
+    static readonly PRICE_CODE_REGULAR = new RegularPricingStrategy()
```
Comment (nicksan222) [inline src/engine/store.ts:12]: Although this is considered a bad practice—since the `Price` class should ideally handle its own types, and the user should instantiate the pricing strategies objects directly—I’ve kept it and marked it as deprecated, so external usage remains unchanged and we avoid breaking changes.

Would be nice in the future to migrate away from this approach!

File: src/engine/movie.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { PricingStrategy } from "./pricing";
+
+export class Movie {
+    constructor(public readonly title: string, public readonly pricingStrategy: PricingStrategy) { }
```
Comment (nicksan222) [inline src/engine/movie.ts:4]: Even though a pure refactor typically means no changes to external behavior, the parameter was renamed because it’s only used internally in the files affected by the refactor. 

This means that external usage remains unchanged, and the tests were simply updated to reflect the new parameter name.

The old parameter was `public readonly priceCode: PriceCode`

File: src/engine/pricing/children.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        if (daysRented > this.baseDays) {
+            // Extended rental: charge only for days beyond the base period
+            // Example: 5 days = (5-3) * $1.50 = $3.00
+            amount += (daysRented - this.baseDays) * this.extraDayRate
+        } else {
+            // Short rental (1-3 days): flat base price regardless of exact days
+            // Example: 1 day = $1.50, 2 days = $1.50, 3 days = $1.50
+            amount += this.basePrice
+        }
+
```
Comment (nicksan222) [inline src/engine/pricing/children.ts:49]: The existing test case already verifies this specific behavior, so the outcome remains unchanged. However, the function has been refactored to improve readability.

That said, the current logic might indicate a potential bug: if a customer exceeds the base days, they still end up paying the same amount on the following day, which may not be the intended behavior.

File: src/engine/pricing/children.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        if (daysRented > this.baseDays) {
+            // Extended rental: charge only for days beyond the base period
+            // Example: 5 days = (5-3) * $1.50 = $3.00
+            amount += (daysRented - this.baseDays) * this.extraDayRate
+        } else {
+            // Short rental (1-3 days): flat base price regardless of exact days
+            // Example: 1 day = $1.50, 2 days = $1.50, 3 days = $1.50
+            amount += this.basePrice
+        }
```
Comment (nicksan222) [inline src/engine/pricing/children.ts:48]: The existing test case already verifies this specific behavior, so the outcome remains unchanged. However, the function has been refactored to improve readability.

That said, the current logic might indicate a potential bug: if a customer exceeds the base days, they still end up paying the same amount on the following day, which may not be the intended behavior.

File: src/engine/store.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    
+    /** @deprecated Use `new ChildrensPricingStrategy()` instead */
+    static readonly PRICE_CODE_CHILDREN = new ChildrensPricingStrategy()
+    
+    /** @deprecated Use `new NewReleasePricingStrategy()` instead */
+    static readonly PRICE_CODE_NEW_RELEASE = new NewReleasePricingStrategy()
```
Comment (nicksan222) [inline src/engine/store.ts:18]: Although this is considered a bad practice—since the `Price` class should ideally handle its own types, and the user should instantiate the pricing strategies objects directly—I’ve kept it and marked it as deprecated, so external usage remains unchanged and we avoid breaking changes.

Would be nice in the future to migrate away from this approach!
